server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -197,12 +265,116 @@ public void validate(final RelativeByteSizeValue value, final Map<Setting<?>, Ob
 
         @Override
         public Iterator<Setting<?>> settings() {
-            final List<Setting<?>> settings = List.of(
-                CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING,
-                CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING,
-                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING
+            return WATERMARK_VALIDATOR_SETTINGS_LIST.iterator();
+        }
+
+    }
+
+    /**
+     * Validates that low, high and flood stage max headrooms adhere to the comparison: flood &lt; high &lt; low.
+     * Also validates that if the low max headroom is set, then the high max headroom must be set as well.
+     * Also validates that if the high max headroom is set, then the flood stage max headroom must be set as well.
+     * Also validates that if max headrooms are set, the respective watermark values should be ratios/percentages.
+     * Else, throws an exception.
+     */
+    static class MaxHeadroomValidator implements Setting.Validator<ByteSizeValue> {
+
+        @Override
+        public void validate(ByteSizeValue value) {
+
+        }
+
+        @Override
+        public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+            final ByteSizeValue lowHeadroom = (ByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_MAX_HEADROOM_SETTING);
+            final ByteSizeValue highHeadroom = (ByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_MAX_HEADROOM_SETTING);
+            final ByteSizeValue floodHeadroom = (ByteSizeValue) settings.get(
+                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_MAX_HEADROOM_SETTING
+            );
+            final ByteSizeValue frozenFloodHeadroom = (ByteSizeValue) settings.get(
+                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_MAX_HEADROOM_SETTING
+            );
+
+            // Ensure that if max headroom values are set, then watermark values are ratios/percentages.
+            final RelativeByteSizeValue low = (RelativeByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING);
+            if (low.isAbsolute()
+                && (lowHeadroom.equals(ByteSizeValue.MINUS_ONE) == false
+                    || highHeadroom.equals(ByteSizeValue.MINUS_ONE) == false
+                    || highHeadroom.equals(ByteSizeValue.MINUS_ONE) == false)) {
```

Comment:
This looks wrong? Is there perhaps a missing test for this?

```suggestion
                    || floodHeadroom.equals(ByteSizeValue.MINUS_ONE) == false)) {
```